xmtp_mls/IDENTITY.md

```diff
@@ -24,59 +24,143 @@ Using per-installation keys provides the following benefits:
+
+Users may revoke an installation as follows:
+
+1. Enumerate active installations by querying for all identity updates under the account. The user may identify each installation by the creation time as well as the installation public key of the credential.
+1. Select the installation to revoke.
+1. The app prompts the user to sign the revocation with their Ethereum wallet. The user is expected to inspect the text and reject the signing request if the data is invalid, for example if the displayed time is incorrect. The format for version 1 of the association text is as follows:
+
+   ```
+   XMTP: Revoke Messaging Access
+
+   Current Time: <ISO 8601 date and time with local UTC offset>
+   Installation ID: <hex(last_20_bytes(keccak256(Ed25519PublicKey)))>
+   ```
+
+1. The signature and related data is then protobuf-serialized to form the revocation:
+
+   ```
+   struct {
+       installation_public_key: bytes,
+       eip191_association: Eip191Association
+   } InstallationRevocation;
+   ```
+
+1. The app publishes the revocation to the server. The server will provide all identity updates (registrations and revocations) for a given wallet address to any client that requests it.
+
+Validation of revocations is identical to the process described in [Credential Validation](#credential-validation), except that a label of `Revoke Messaging Access` is used.
+
+Once an installation is revoked, it cannot be re-registered or re-provisioned. The time displayed on the revocation is for informational purposes only.
+
+### Installation synchronization
+
+At any time in the course of a conversation, the list of valid installations for the participating wallet addresses may change via registration or revocation.
+
+Clients must perform the following validation prior to publishing each payload on the group:
+
+1. Assemble a list of wallet addresses in the conversation from the leaf nodes.
+1. Fetch all identity updates on those wallet addresses.
+1. Validate the credentials and revocations and construct a list of valid installations (registered installations minus revoked installations).
+1. Publish a commit to remove nodes from the conversation that are not in the list of valid installations.
+1. Publish a commit to add nodes to the conversation that are in the list of valid installations and not already present.
+
+XMTP clients may perform performance optimizations, such as caching installation lists with a short TTL.
 
-**Installation management**
+To support revocations (Q2+ 2024), clients must additionally perform the following validation on receiving each payload in the group:
 
-At any time, the user may enumerate active installations by querying for all identity updates under the account. The user may identify each installation by the creation time as well as the installation key from the signing text.
+1. Query the server for identity updates on the sending wallet address and verify that the sending installation has not been revoked.
```

Comment:
I'm not sure this is going to be viable as part of the message processing flow. Any race condition here could lead to part of the group accepting a commit and part of a group rejecting it. Even if we are only checking whether the installation was revoked at the time the message was sent, and not the time it was read, it still opens the door to potential race conditions given that revocation state does not have the same total ordering properties as topic contents.

I propose we don't check for revocation when reading payloads and instead do it before synchronizing the group state. Anyone with privileges allowing them to remove users would check the network for any revoked members and submit a commit to remove the revoked installation from the group before processing any other intents.

How this interacts with group permissions is TBD. Maybe we want a carve-out allowing anyone to remove a revoked installation so that we don't have to wait for an admin to come online. Maybe we have each user go through all their groups and remove themselves as part of revocation.

Reply:
My preference is to have total ordering between revocations and commits, which is a little more robust, but I concede that the current direction of decentralization work doesn't seem to be heading that way.

Updated the revocation approach in https://github.com/xmtp/libxmtp/pull/353/commits/a54dbad33ed3fad21565dc18381a65917159c0d0!